### PR TITLE
role: Fix role's hash_params

### LINKED ISSUE #20596: Variables not reflecting values in playbook when using with_items
I have not found a similar issue open already, however I must say I did not verify all 1505.
The problem exists in 2.2.0.0, 2.2.1.0 and when using a checkout of the devel branch.

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
with_items only updates variables once if the type is the same

##### ANSIBLE VERSION

```
ansible 2.2.1.0
  config file = /Users/evert/git/ansible-variable-bug/ansible.cfg
  configured module search path = ['./library']
```

##### CONFIGURATION

Ansible.cfg
```
[defaults]
hostfile = hosts
roles_path = roles
library = ./library
retry_files_enabled = False
```

##### OS / ENVIRONMENT

Macintosh, Ubuntu 14

##### SUMMARY

This is either a bug in Ansible or I don’t see my error. The example playbook launches a role 2 times, first with testvalue = true, testvalue2=false. Afterwards with testvalue = false, testvalue2 = true. When using or printing out the variables, it prints true - false, true - false twice instead of using the adapted variables.

The values are changed if the type changes. For example by first using a bool testvalue = true and afterwards a string (between quotes) testvalue = "false"

##### STEPS TO REPRODUCE

A simplified example can be found here:
https://github.com/EvertMDC/ansible-variable-bug

```yaml
# does not update the variables
ansible-playbook testbug.yml 
# does update the variables, first using boolean and afterwards strings to set the values
ansible-playbook testbug_different_vartype.yml 
```

##### EXPECTED RESULTS

Printed variables should reflect those given in the playbook

##### ACTUAL RESULTS

Printed variables are not updated if the type remains the same.

#### Not updated, output of testbug.yml
You will see true - false, true - false
```
PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [role_testbug : Printing out values] **************************************
ok: [localhost] => (item={u'k': u'testvalue', u'v': u'true'}) => {
    "item": {
        "k": "testvalue",
        "v": "true"
    },
    "msg": "value true key testvalue"
}
ok: [localhost] => (item={u'k': u'testvalue2', u'v': u'false'}) => {
    "item": {
        "k": "testvalue2",
        "v": "false"
    },
    "msg": "value false key testvalue2"
}

TASK [role_testbug : Printing out values] **************************************
ok: [localhost] => (item={u'k': u'testvalue', u'v': u'true'}) => {
    "item": {
        "k": "testvalue",
        "v": "true"
    },
    "msg": "value true key testvalue"
}
ok: [localhost] => (item={u'k': u'testvalue2', u'v': u'false'}) => {
    "item": {
        "k": "testvalue2",
        "v": "false"
    },
    "msg": "value false key testvalue2"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0
```
#### Updated, output of testbug_different_vartype.yml
You will see true - false, false - true
```
PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [role_testbug : Printing out values] **************************************
ok: [localhost] => (item={u'k': u'testvalue', u'v': u'true'}) => {
    "item": {
        "k": "testvalue",
        "v": "true"
    },
    "msg": "value true key testvalue"
}
ok: [localhost] => (item={u'k': u'testvalue2', u'v': u'false'}) => {
    "item": {
        "k": "testvalue2",
        "v": "false"
    },
    "msg": "value false key testvalue2"
}

TASK [role_testbug : Printing out values] **************************************
ok: [localhost] => (item={u'k': u'testvalue', u'v': u'false'}) => {
    "item": {
        "k": "testvalue",
        "v": "false"
    },
    "msg": "value false key testvalue"
}
ok: [localhost] => (item={u'k': u'testvalue2', u'v': u'true'}) => {
    "item": {
        "k": "testvalue2",
        "v": "true"
    },
    "msg": "value true key testvalue2"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0
```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Fix based upon work done by alikins.

Fixes: #20596

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
changelogs/fragments/20596-role-param_fix.yaml
lib/ansible/playbook/role/__init__.py
test/units/playbook/role/test_role.py

### CODE REVIEW

File: lib/ansible/playbook/role/__init__.py [hunk 1]
```diff
@@ -59,6 +59,7 @@ def hash_params(params):
     # Any container is unhashable if it contains unhashable items (for
     # instance, tuple() is a Hashable subclass but if it contains a dict, it
     # cannot be hashed)
+    new_params = None
```
Comment (bcoca) [inline lib/ansible/playbook/role/__init__.py]: isn't this useless ince we we do `= set()` on each case below?

**Comment by Akasurde** [comment]
@bcoca Could you please review this ? Thanks.

**Comment by Akasurde** [comment]
ping @bcoca

File: lib/ansible/playbook/role/__init__.py [hunk 2]
```diff
@@ -68,7 +68,7 @@ def hash_params(params):
                 new_params = set()
                 for k, v in params.items():
                     # Hash each entry individually
-                    new_params.update((k, hash_params(v)))
+                    new_params.add((k, hash_params(v)))
```
Comment (bcoca) [inline lib/ansible/playbook/role/__init__.py:71]: ~won't this give different behaviour compared to `frozenset` above if `.items()` is non unique?~ nvmd, frozenset below ... which makes me think this change is either a noop or it is changing which k value survives?

update will add each element of the tuple to the set, while add adds the tuple itself to the set, so this does seem like a correct change.

**Review by bcoca — APPROVED** [review]

**Pull request merged**